eth_abi/registry.py

```diff
@@ -0,0 +1,94 @@
+from .utils.parsing import collapse_type
+
+
+class ABIRegistryError(Exception):
+    pass
+
+
+class NoEntryFound(ABIRegistryError):
+    pass
+
+
+class PredicateMapping:
+    def __init__(self):
+        self._values = {}
+        self._labeled_predicates = {}
+
+    def add(self, predicate, value, label=None):
+        if label is not None:
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
+
+    def remove_by_predicate(self, predicate):
+        # Delete the predicate mapping to the previously stored value
+        try:
+            del self._values[predicate]
+        except KeyError:
+            raise KeyError('Predicate not found')
+
+        # Delete any label which refers to this predicate
+        try:
+            label = self._label_for_predicate(predicate)
+        except ValueError:
+            pass
+        else:
+            del self._labeled_predicates[label]
+
+    def _label_for_predicate(self, predicate):
+        for key, value in self._labeled_predicates.items():
+            if value is predicate:
+                return key
+
+        raise ValueError('Predicate not referred to by a label')
+
+    def remove_by_label(self, label):
+        try:
+            predicate = self._labeled_predicates[label]
+        except KeyError:
+            raise KeyError('Label not found')
+
+        del self._labeled_predicates[label]
+        del self._values[predicate]
+
+
+class ABIRegistry:
+    def __init__(self):
+        self._encoders = PredicateMapping()
+
+    def register_encoder(self, lookup, encoder, label=None):
+        """
+        Registers an encoder under the given type lookup and optional label.
+        """
+        if callable(lookup):
+            self._encoders.add(lookup, encoder, label)
+            return
+
+        if isinstance(lookup, str):
```

Comment:
This is a little more awkward with the predicate approach, but probably still worth doing: 

We talked about checking the `str` against existing `str` predicates at register-time. Unfortunately, all the other checks seem likely to only trigger at encode time.

Reply:
Need some clarification on this.  Are you wondering how we'll detect conflicts at register time if everything is a predicate?

Reply:
Actually, I missed that this was adding the string `lookup` as a label. So resolving the other comment will address this issue.

Reply:
I was just wanting to make sure that registering encoders twice with the same string would fail.